### PR TITLE
Add alternative model serializations for inference

### PR DESCRIPTION
**Pull request by drubinstein**
This PR is a pretty large refactor. One issue we've seen is that not all users are TensorFlow users and TensorFlow is a pretty heavy dependency. To alleviate that, this PR adds in TFLite, CoreML and ONNX serialized versions of the models. PyTorch is not included for now, but we plan to include it in the future.

Instead of TF being a default dependency, depending on the platform, a different model serialization format will be used by defualt. Instead TF is now an extra that can be installed via `basic-pitch[tf]`.

- On Linux, TFLite will be used via the `tflite-runtime` package (which only supports linux)
- On Mac, CoreML will be used via the `coremltools` package (which only supports mac)
- On Windows, ONNX will be used via the `onnxruntime` package (which supports all platforms)

However, if TF is detected on the system, it will be used instead. Ideally, this change will speed up initialization and install times at the cost of a slightly larger binary due to the inclusion of all the new serialized models. 

In addition, related to a red herring when failing to install scipy in Python 3.11, I decided to also move from `setup.cfg` to `pyproject.toml` format in this PR. I can move that to another PR if desired.

I believe we can call this a minor bump since there are no _breaking_ changes, but we can also call this a major bump due to the possible results changes for users.

### CODE REVIEW

File: .github/workflows/tox.yml [hunk 1]
```diff
@@ -6,30 +6,34 @@ on:
...
           - "3.8"
+        include:
+          - os: macos-latest-xlarge
+            py: "3.10.11"
+          - os: macos-latest-xlarge
+            py: "3.11.8"
     steps:
       - name: Setup python for test ${{ matrix.py }}
-        uses: actions/setup-python@v3
+        uses: actions/setup-python@v4
```
Comment (bgenchel) [inline .github/workflows/tox.yml:28]: where does this v4 come from?

File: README.md [hunk 2]
```diff
@@ -44,14 +44,31 @@ To update Basic Pitch to the latest version, add `--upgrade` to the above comman
...
 #### Compatible Environments:
 - MacOS, Windows and Ubuntu operating systems
-- Python versions 3.7, 3.8, 3.9, 3.10
+- Python versions 3.7, 3.8, 3.9, 3.10, 3.11
 - **For Mac M1 hardware, we currently only support python version 3.10. Otherwise, we suggest using a virtual machine.**
 
 
+### Model Runtime
+
+Basic Pitch comes with the original TensorFlow model and the TensorFlow model converted to [CoreML](https://developer.apple.com/documentation/coreml), [TensorFlowLite](https://www.tensorflow.org/lite), and [ONNX](https://onnx.ai/). By default, Basic Pitch will _not_ install TensorFlow as a dependency *unless you are using Python>=3.11*. Instead, by default, CoreML will be installed on MacOS, TensorFlowLite will be installed on Linux and ONNX will be installed on Windows. If you want to install TensorFlow along with the default model inference runtime, you can install TensorFlow via `pip install basic-pitch[tf]`.
```
Comment (bgenchel) [inline README.md:53]: why is tensorflowlite mapped to linux and onnx mapped to windows?

File: README.md [hunk 3]
```diff
@@ -44,14 +44,31 @@ To update Basic Pitch to the latest version, add `--upgrade` to the above comman
...
+By default, Basic Pitch will attempt to load a model in the following order:
+
+1. TensorFlow
+2. CoreML
+3. TensorFlowLite
+4. ONNX
+
+Additionally, the module variable ICASSP_2022_MODEL_PATH will default to the first available version in the list.
+
+We will explain how to override this priority list below. Because all other model serializations were converted from TensorFlow, we recommend using TensorFlow when possible.
```
Comment (bgenchel) [inline README.md:70]: the advice on 70 seems to conflict a bit with the above; if it is recommended shouldn't it always be the default?

File: basic_pitch/__init__.py [hunk 4]
```diff
@@ -15,13 +15,81 @@
...
+import enum
+import logging
 import pathlib
 
-__author__ = "Spotify"
-__version__ = "0.3.0"
-__email__ = "basic-pitch@spotify.com"
-__demowebsite__ = "https://basicpitch.io"
-__description__ = "Basic Pitch, a lightweight yet powerful audio-to-MIDI converter with pitch bend detection."
-__url__ = "https://github.com/spotify/basic-pitch"
```
Comment (bgenchel) [inline basic_pitch/__init__.py:22]: doesn't seem like we needed to get rid of this in adding the new stuff. What's the reason for the removal?

File: basic_pitch/inference.py [hunk 5]
```diff
@@ -18,12 +18,37 @@
...
+        import tensorflow.lite as tflite
+
+try:
+    import onnxruntime as ort
+except ImportError:
+    pass
 
-from tensorflow import Tensor, signal, keras, saved_model
 import numpy as np
+import numpy.typing as npt
```
Comment (bgenchel) [inline basic_pitch/inference.py:51]: is it correct style to have a block of code running in between the imports?

File: basic_pitch/inference.py [hunk 6]
```diff
@@ -33,16 +58,132 @@
...
+                self.model_type = Model.MODEL_TYPES.COREML
+                self.model = ct.models.MLModel(str(model_path))
+                return
+            except Exception as e:
+                if str(model_path).endswith(".mlpackage"):
+                    logging.warning(
+                        "Could not load CoreML file %s even "
+                        "though it looks like a CoreML file with error %s. "
+                        "Are you sure it's a TFLite file?",
+                        model_path,
```
Comment (bgenchel) [inline basic_pitch/inference.py:107]: I think you mean CoreML file here

File: basic_pitch/inference.py [hunk 7]
```diff
@@ -90,11 +230,15 @@ def get_audio_input(
 
     original_length = audio_original.shape[0]
     audio_original = np.concatenate([np.zeros((int(overlap_len / 2),), dtype=np.float32), audio_original])
-    audio_windowed, window_times = window_audio_file(audio_original, hop_size)
-    return audio_windowed, window_times, original_length
+    for window, window_time in window_audio_file(audio_original, hop_size):
+        yield np.expand_dims(window, axis=0), window_time, original_length
```
Comment (bgenchel) [inline basic_pitch/inference.py:234]: what is the rationale behind the switch to yield over return here?

File: basic_pitch/saved_models/icassp_2022/nmp.mlpackage/Manifest.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        },
+        "C784FB56-1BA8-4A0E-AC2A-F0388FE1A2DB": {
+            "author": "com.apple.CoreML",
+            "description": "CoreML Model Weights",
+            "name": "weights",
+            "path": "com.apple.CoreML/weights"
+        }
+    },
+    "rootModelIdentifier": "7EB69391-53BD-4281-BFA1-831180579ABC"
+}
```
Comment (bgenchel) [inline basic_pitch/saved_models/icassp_2022/nmp.mlpackage/Manifest.json:18]: what is the purpose of this file, or where does this way of specifying come from?

**Review by bgenchel — CHANGES_REQUESTED** [review]
Most things look good from what I can understand though i'd like to understand a bit more before approving; there is one comment I left for a small change in the __init__.py text output.

File: .github/workflows/tox.yml (hunk 1 above)
Comment (drubinstein) [reply .github/workflows/tox.yml:28]: It was at the time, the most recent version of the [github-built-in setup-python action](https://github.com/actions/setup-python)

File: README.md (hunk 2 above)
Comment (drubinstein) [reply README.md:53]: tflite-runtime only provides [linux whls](https://pypi.org/project/tflite-runtime/#files). Onnx at least provides windows wheels. To use TFLite on windows, you need to TF. Since BP was tensorflow, first, I'd rather stick to tflite when possible.

File: README.md (hunk 3 above)
Comment (drubinstein) [reply README.md:70]: It's recommended if you're already installing tensorflow, however, tensorflow is quite heavy so I dont recommend it being the default (which is why I wrote when possible). I can try to rephrase

File: basic_pitch/__init__.py (hunk 4 above)
Comment (drubinstein) [reply basic_pitch/__init__.py:22]: Not really needed. It's all in pyproject.toml

File: basic_pitch/inference.py (hunk 5 above)
Comment (drubinstein) [reply basic_pitch/inference.py:51]: It isn't good style, but there's not really a better solution I could come up with. Generally I've seen this pattern done when you're dealing with extras that may or may not be installed.

File: basic_pitch/inference.py (hunk 6 above)
Comment (drubinstein) [reply basic_pitch/inference.py:107]: I did. Will fix.

File: basic_pitch/inference.py (hunk 7 above)
Comment (drubinstein) [reply basic_pitch/inference.py:234]: Instead of returning a list of items and allocating all the memory for that, yield lets us return just whats needed when its needed.

File: basic_pitch/saved_models/icassp_2022/nmp.mlpackage/Manifest.json (hunk 8 above)
Comment (drubinstein) [reply basic_pitch/saved_models/icassp_2022/nmp.mlpackage/Manifest.json:18]: It's a coreml file. `.mlpackage` is the extension for coreml files.

**Review by bgenchel — APPROVED** [review]

File: tox.ini [hunk 9]
```diff
@@ -11,7 +11,7 @@ requires =
 [testenv]
 deps = -e .[test]
 commands =
-    pytest tests --ignore tests/test_nn.py -s {posargs}
+    pytest tests --ignore tests/test_nn.py {posargs}
```
Comment (bgenchel) [inline tox.ini:14]: is this to speed things up? or, what's the reason for this change?

Comment (drubinstein) [reply tox.ini:14]: `-s` is the pytest argument for not capturing stdout. It's not really needed so I removed it.

File: tests/test_inference.py [hunk 10]
```diff
@@ -55,12 +55,11 @@ def test_predict() -> None:
     assert all(note_pitch_max)
     assert isinstance(note_events, list)
 
-    expected_model_output = np.load("tests/resources/vocadito_10/model_output.npz")
+    expected_model_output = np.load("tests/resources/vocadito_10/model_output.npz", allow_pickle=True)["arr_0"].item()
```
Comment (bgenchel) [inline tests/test_inference.py:58]: could you explain what's happening here a bit? I guessed from above that you were forcing a lowered version of tensorflow in order to avoid any issues with our current serialization, but this line makes it seem that you might be adapting to the new serialization format? If that makes sense.

Comment (drubinstein) [reply tests/test_inference.py:58]: I had to remake all the test data and this is what made it work. The new format is in the keras3.0 / model loading layer. Not in numpy. There may have been some changes to the numpy format along the way.

**Review by bgenchel — APPROVED** [review]
